dm-dedup-rw.c

```diff
@@ -64,15 +66,19 @@ static int merge_data(struct dedup_config *dc, struct page *page,
 		goto out;
 	}
 
-	src_page_vaddr = page_address(bio->bi_io_vec->bv_page);
-	dest_page_vaddr = page_address(page);
-
-	src_page_vaddr = src_page_vaddr + bio->bi_io_vec->bv_offset;
 	/* Locating the right sector to merge */
-	dest_page_vaddr = dest_page_vaddr + (to_bytes(position));
+	dest_page_vaddr = page_address(page) + to_bytes(position);
+
+	bio_for_each_segment(bvec, bio, iter) {
+		src_page_vaddr = bio_data(bio);
+
+		/* Merging Data */
+		memmove(dest_page_vaddr, src_page_vaddr, bio_cur_bytes(bio));
+
+		/* Updating destination address */
+		dest_page_vaddr += bio_cur_bytes(bio);
+    }
```

Comment:
The bio_for_each_segment macro looks like this

> #define bio_for_each_segment(bvl, bio, iter)
>     __bio_for_each_segment(bvl, bio, iter, (bio)->bi_iter)

The initialization argument in the for loop in __bio_for_each_segment macro is

> #define __bio_for_each_segment(bvl, bio, iter, start)
>     for (iter = (start);

This is how the macro is written. I dont know why they are getting a new iter instead of just getting the pointer.